MAGN-4745 Colour the nodes according to the delta state

ramramps opened the pull request:
This PR shows many changes because of the revert and merge. 

The changes are same as the PR : https://github.com/DynamoDS/Dynamo/pull/3364
with few changes that controls the showrunpreview behavior. 

This PR had the fix for custom nodes as well. 

- [ ] @ikeough

ikeough commented:
@ramramps We're getting there. I would like to move the properties that are used to control the visibility of run preview into the view layer because, as far as I can tell, they don't have any purpose in the model layer.

ramramps commented:
@ikeough 

1. HomeWorkspaceModel should know about ShowRunPreview., inorder to get the next executing nodes. Moved the ShowRunPreview to Workspacemodel. So the DyanmoViewModel will get the value from Workspacemodel everytime and update the UI. 
2. ShowExecutionPreview lives in nodemodel. But the function that sets the ShowExectionPreview is moved to Nodemodel constructor. I removed it from HomeWorkspacemodel. 
3.When ShowRunPreview value is changed to true, it looks for the next executing nodes. If it is false, it just changes the Execution preview to false. 

I checked the workflow and arrived at this solution. If this doesn't make sense or if I had missed something let me know.

ikeough commented on src/DynamoCore/Models/HomeWorkspaceModel.cs:
```diff
@@ -297,6 +323,52 @@ protected virtual void OnEvaluationCompleted(EvaluationCompletedEventArgs e)
             if (handler != null) handler(this, e);
         }
 
+        public override void GetExecutingNodes()
```
The name of this method makes me believe I'm going to get a set of nodes returned to me. But this is not the case. Instead, we return void and within the method we modify a bunch of properties on some nodes. I would prefer that this method returned a set of nodes. Then, you could call this method from the `WorkspaceViewModel`, and set the `ShowExecutionPreview` property on all the `NodeViewModels`. 

I agree that this method should continue to live in the model layer, but we should call it and affect changes to the UI from the view model layer and above.

ikeough commented on src/DynamoCore/Models/NodeModel.cs:
```diff
@@ -1434,6 +1435,11 @@ protected override void DeserializeCore(XmlElement nodeElement, SaveContext cont
         #endregion
 
         #region Dirty Management
+        //TODO: Refactor Property into Automatic with private(?) setter
```
Are these TODOs still open? Are they associated with a YT task?

ikeough commented on src/DynamoCore/Models/NodeModel.cs:
```diff
@@ -1476,6 +1482,32 @@ protected virtual ExecutionHints GetExecutionHintsCore()
         }
 
+
+        public bool ShowExecutionPreview
```
I want `ShowExecutionPreview` in the view model layer. This property has no logical reason for being in the model.

ikeough commented on src/DynamoCore/Models/WorkspaceModel.cs:
```diff
@@ -436,6 +437,19 @@ public bool RunEnabled
+
+        private bool showRunPreview;
+
+        public bool ShowRunPreview
```
Similar to above, this property should live in the view model. It doesn't make sense in the model.

ikeough commented on src/DynamoCore/Models/WorkspaceModel.cs:
```diff
@@ -606,6 +621,14 @@ public virtual void OnNodesModified()
+        /// <summary>
+        /// This gets the next set of executing nodes 
+        /// </summary>
+        public virtual void GetExecutingNodes()
```
If this is never going to have a body, can we make it abstract?

ikeough commented:
@ramramps I might be missing something here, but what I am imaging for this flow is:

Enable "Show Run Preview"

----- View / View Model Boundary -----

Binding set ShowRunPreview
On Node property changes and node add/delete...

----- ViewModel / Model Boundary -----

GetExecutingNodes

----- Model / ViewModel Boundary -----

Set ShowExecutionPreview on NodeViewModels

----- ViewModel / View Boundary -----

Bindings update UI

ramramps commented:
@ikeough - I made those changes what we discussed today. Now ShowExecutionPreview lives only NodeViewModel (So,the connectorViewModel is changed a bit to get the ShowExecuctionPreview from NodeViewModel instead of a NodeModel). Also, made changes related to RunSettings. 
ShowRunPreview  lives only in DyanmoViewModel. I believe , I had removed the dependencies around the models now and those parameters lives only in viewmodels. Please let me know if this doesnt work or if I had to more changes around this. I am happy to do the changes :)

ikeough commented:
@ramramps Thanks for the changes. LGTM, when you clear the merge conflict.

ramramps commented:
@ikeough  - Ian, I forgot to update the tooltip yesterday Just got to see this after the merge. 
If this looks OK, I can merge it.

ikeough commented:
@ramramps That's fine.

ramramps commented:
@ikeough - Shall I merge this branch? I dont want to break something in the middle
